### PR TITLE
[TOPI] Improve schedule for injective on x86

### PR DESCRIPTION
**Pull request by merrymercy**
The existing schedule does not handle parallel correctly for all shapes.
For example, it does not do parallel for shape (1, 4096, 1024) at all.
This pr makes the parallel strategy more general.

### CODE REVIEW

File: topi/include/topi/x86/injective.h [hunk 1]
```diff
@@ -44,14 +47,29 @@ namespace x86 {
...
+  int64_t prod_len = 1;
+
+  for (const auto& axis : axes) {
+    const auto* pint = axis->dom->extent.as<IntImmNode>();
+    if (pint == nullptr) {
+      break;
+    }
+    prod_len *= pint->value;
+    to_fuse.push_back(axis);
+    if (prod_len >= tvm::runtime::threading::MaxConcurrency()) {
```
Comment (FrozenGene) [inline topi/include/topi/x86/injective.h:61]: According to build CI information, we should change the `visibility` of `tvm::runtime::threading::MaxConcurrency` according to append `TVM_DLL`

**Comment by merrymercy** [comment]
I pushed some changes to my branch. But why does not GitHub sync the changes to this PR?

File: topi/include/topi/x86/injective.h [hunk 2]
```diff
@@ -44,14 +47,29 @@ namespace x86 {
  * \return The updated schedule.
  */
 inline Schedule schedule_injective_from_existing(Schedule sch, const Tensor& out) {
-  auto axis = sch[out]->op.as<ComputeOpNode>()->axis;
-  if (axis.size() == 4) {
-    auto n = axis[0];
-    auto c = axis[1];
-    auto fused = detail::Fuse(sch[out], {n, c});  // for nhwc layout, fuse n and h
-    sch[out].parallel(fused);
+  const auto& axes = sch[out]->op.as<ComputeOpNode>()->axis;
```
Comment (kevinthesun) [inline topi/include/topi/x86/injective.h:50]: Is this schedule func used in the common compilation flow? In op strategy we use python version of injective schedule and it should handle parallelism appropriately.

**Comment by icemelon9** [comment]
Could you also update the injective schedule in the python as well?

**Comment by tqchen** [comment]
cc @merrymercy please followup

**Comment by merrymercy** [comment]
Why do we have both c++ version and python version? Should we just delete the python version?

**Comment by merrymercy** [comment]
I added a python version. It handles parallel and vectorization generally for most input shapes.
I tested in with BERT (seq_len=128, batch=1) on c5.9xlarge. It reduces the latency from 19.0ms to 18.70ms.

My questions:
1. Why do we have both c++ version and python version? Should we just delete the python version?

**Comment by tqchen** [comment]
If c++ version is good we can call into the c++ version from the pyone

**Comment by merrymercy** [comment]
Currently, both c++ and python versions work well. This PR is ready to be merged from my side.
I have no idea how to solve the mess of duplicated python and c++ topi functions.

**Comment by icemelon9** [comment]
@merrymercy Yes, I agree with @tqchen. If c++ version works well, we could just remove the python implementation and use c++ one instead.

**Comment by icemelon9** [comment]
Probably also worth checking with @kevinthesun. I remember last time he saw some performance regression due to the injective schedule.

**Comment by kevinthesun** [comment]
Yes. If we haven't done so, I suggest testing at least resnet50  to make sure no any performance regression occurs.

**Comment by tqchen** [comment]
@merrymercy please follow up

**Comment by merrymercy** [comment]
I close this PR because after more test I found the current python version in tvm master works well. This pr is trying to fix the c++ version, which is only used in the old codebase.

**Pull request closed without merging**